Cython/Compiler/ModuleNode.py

```diff
@@ -2855,15 +2855,11 @@ def generate_module_state_start(self, env, code):
         code.putln('PyObject *%s;' % Naming.empty_unicode)
         if Options.pre_import is not None:
             code.putln('PyObject *%s;' % Naming.preimport_cname)
-        for type_cname, used_name in Naming.used_types_and_macros:
-            code.putln('#ifdef %s' % used_name)
-            code.putln('PyTypeObject *%s;' % type_cname)
-            code.putln('#endif')
 
     def generate_module_state_end(self, env, modules, globalstate):
         module_state = globalstate['module_state_end']
-        module_state_clear = globalstate['module_state_clear']
-        module_state_traverse = globalstate['module_state_traverse']
+        module_state_clear = globalstate['module_state_clear_end']
+        module_state_traverse = globalstate['module_state_traverse_end']
```

Comment:
I think we might want to put the type cleanup near the end rather than the beginning. Although that's probably less relevant given that objects own a reference to their types now. But if I were to chose, I'd not take the risk.

Reply:
The order should be:
* `module_state_clear` - most of the `ModuleNode` generated code for clear
* `module_state_clear_contexts` - utility code injected into the clear function
* `module_state_clear_end` - pretty much just `return 0; }`

So the types are put near the end as you suggest.